streaming/stream_transfer_task.cc

```diff
@@ -237,18 +237,23 @@ future<> stream_transfer_task::execute() {
     }).then([this, id, plan_id] {
         _mutation_done_sent = true;
         sslog.debug("[Stream #{}] GOT STREAM_MUTATION_DONE Reply from {}", plan_id, id.addr);
-    }).handle_exception([this, plan_id, cf_id, id] (std::exception_ptr ep) {
+    }).handle_exception_type([this, plan_id, cf_id, id] (const replica::no_such_column_family& e) {
         // If the table is dropped during streaming, we can ignore the
         // errors and make the stream successful. This allows user to
         // drop tables during node operations like decommission or
         // bootstrap.
-        if (!session->manager().db().column_family_exists(cf_id)) {
-            sslog.warn("[Stream #{}] Ignore the table with table_id {} which is dropped during streaming: {}", plan_id, cf_id, ep);
-            if (!_mutation_done_sent) {
-                return session->manager().ms().send_stream_mutation_done(id, plan_id, _ranges, cf_id, session->dst_cpu_id);
-            }
+        //
+        // the db table metadata on different shards are not necessarily in
+        // sync, but if the table is dropped on any one of them, the exception
+        // is thrown. so we don't check the db on _current_ shard, whose db
+        // table metadata could be waiting for a write lock at this moment.
+        assert(cf_id == std::get<table_id>(e.table()));
```

Comment:
What if one uses

no_such_column_family::no_such_column_family(std::string_view ks_name, std::string_view cf_name)

to throw the exception?

The table id will not be present.

Reply:
i was trying to be paranoid, but not too paranoid. if this exception is thrown, it must be thrown by `auto tbl = sm.db().find_column_family(cf_id).shared_from_this();` at line 214.  actually, i would want to drop this `assert()` if we are aiming at a fix proposed by #17028, as it depends on a change in `no_such_column_family`, which might not be useful without this change.